packages/plugin-ext/src/hosted/browser/hosted-plugin.ts

```diff
@@ -62,26 +66,43 @@ export class HostedPluginSupport {
     @inject(PreferenceProviderProvider)
     protected readonly preferenceProviderProvider: PreferenceProviderProvider;
 
+    @inject(PreferenceServiceImpl)
+    private readonly preferenceServiceImpl: PreferenceServiceImpl;
+
+    @inject(PluginPathsService)
+    private readonly pluginPathsService: PluginPathsService;
+
+    @inject(StoragePathService)
+    private readonly storagePathService: StoragePathService;
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @inject(MonacoTextmateService)
+    protected readonly monacoTextmateService: MonacoTextmateService;
+
+    @inject(CommandRegistry)
+    protected readonly commands: CommandRegistry;
+
     private theiaReadyPromise: Promise<any>;
-    private frontendExtManagerProxy: PluginManagerExt;
-    private backendExtManagerProxy: PluginManagerExt;
+
+    protected readonly managers: PluginManagerExt[] = [];
 
     // loaded plugins per #id
-    private loadedPlugins: Set<string> = new Set<string>();
+    private readonly loadedPlugins = new Set<string>();
 
-    // per #hostKey
-    private rpc: Map<string, RPCProtocol> = new Map<string, RPCProtocol>();
+    protected readonly activationEvents = new Set<string>();
 
-    constructor(
-        @inject(PreferenceServiceImpl) private readonly preferenceServiceImpl: PreferenceServiceImpl,
-        @inject(PluginPathsService) private readonly pluginPathsService: PluginPathsService,
-        @inject(StoragePathService) private readonly storagePathService: StoragePathService,
-        @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService,
-    ) {
+    @postConstruct()
+    protected init(): void {
         this.theiaReadyPromise = Promise.all([this.preferenceServiceImpl.ready, this.workspaceService.roots]);
-        this.storagePathService.onStoragePathChanged(path => {
-            this.updateStoragePath(path);
-        });
+        this.storagePathService.onStoragePathChanged(path => this.updateStoragePath(path));
+
+        for (const id of this.monacoTextmateService.activatedLanguages) {
+            this.activateByLanguage(id);
+        }
+        this.monacoTextmateService.onDidActivateLanguage(id => this.activateByLanguage(id));
+        this.commands.onWillExecuteCommand(event => this.ensureCommandHandlerRegistration(event));
```

Comment:
With the proposed change, each time I execute a command in Theia, this code will run. Is this the expected behavior? Thanks!

Reply:
It is, all commands which are not contributed by plugin system is filtered out `this.hasCommand` check below. Waiting happens only if a command is contributed by the plugin system, otherwise it should continue immediately.

Reply:
> `this.hasCommand` check below

I have ~checked~ debugged it, that's why I have asked.

Reply:
actually i'm probably wrong, if a command is not contributed by plugin system we should not wait, but we still should fire onCommand activation event!

Reply:
I've updated this code to fire an activation event and fix https://github.com/theia-ide/theia/pull/5622#issuecomment-508076675